src/transfer-manager.ts

```diff
@@ -0,0 +1,342 @@
+ * @experimental
+ */
+const DEFAULT_PARALLEL_CHUNKED_DOWNLOAD_LIMIT = 2;
+/**
+ * The minimum size threshold in bytes at which to apply a chunked download strategy when calling downloadFileInChunks.
+ * @experimental
+ */
+const DOWNLOAD_IN_CHUNKS_FILE_SIZE_THRESHOLD = 256 * 1024 * 1024;
+/**
+ * The chunk size in bytes to use when calling downloadFileInChunks.
+ * @experimental
+ */
+const DOWNLOAD_IN_CHUNKS_DEFAULT_CHUNK_SIZE = 10 * 1024 * 1024;
+const EMPTY_REGEX = '(?:)';
+export interface UploadManyFilesOptions {
+  concurrencyLimit?: number;
+  skipIfExists?: boolean;
+  prefix?: string;
+  passthroughOptions?: Omit<UploadOptions, 'destination'>;
+}
+
+export interface DownloadManyFilesOptions {
+  concurrencyLimit?: number;
+  prefix?: string;
+  stripPrefix?: string;
+  passthroughOptions?: DownloadOptions;
+}
+
+export interface DownloadFileInChunksOptions {
+  concurrencyLimit?: number;
+  chunkSizeBytes?: number;
+  destination?: string;
+}
+
+/**
+ * Create a TransferManager object to perform parallel transfer operations on a Cloud Storage bucket.
+ *
+ * @class
+ * @hideconstructor
+ *
+ * @param {Bucket} bucket A {@link Bucket} instance
+ * @experimental
+ */
+export class TransferManager {
+  bucket: Bucket;
+  constructor(bucket: Bucket) {
+    this.bucket = bucket;
+  }
+
+  /**
+   * @typedef {object} UploadManyFilesOptions
+   * @property {number} [concurrencyLimit] The number of concurrently executing promises
+   * to use when uploading the files.
+   * @property {boolean} [skipIfExists] Do not upload the file if it already exists in
+   * the bucket. This will set the precondition ifGenerationMatch = 0.
+   * @property {string} [prefix] A prefix to append to all of the uploaded files.
+   * @property {object} [passthroughOptions] {@link UploadOptions} Options to be passed through
+   * to each individual upload operation.
+   * @experimental
+   */
+  /**
+   * Upload multiple files in parallel to the bucket. This is a convenience method
+   * that utilizes {@link Bucket#upload} to perform the upload.
+   *
+   * @param {array} [filePaths] An array of fully qualified paths to the files.
+   * to be uploaded to the bucket
+   * @param {UploadManyFilesOptions} [options] Configuration options.
+   * @returns {Promise<UploadResponse[]>}
+   *
+   * @example
+   * ```
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const bucket = storage.bucket('my-bucket');
+   * const transferManager = new TransferManager(bucket);
+   *
+   * //-
+   * // Upload multiple files in parallel.
+   * //-
+   * const response = await transferManager.uploadManyFiles(['/local/path/file1.txt, 'local/path/file2.txt']);
+   * // Your bucket now contains:
+   * // - "local/path/file1.txt" (with the contents of '/local/path/file1.txt')
+   * // - "local/path/file2.txt" (with the contents of '/local/path/file2.txt')
+   * ```
+   * @experimental
+   */
+  async uploadManyFiles(
+    filePaths: string[],
```

Comment:
[Optional] We have an opportunity to get fancy here to really emphasize throughput; what if we allowed iterators and async iterators?

Use cases: walking large directory you want to walk asynchronously (reduces memory usage)

Reply:
We can probably wait until customers ask for this functionality though.

Reply:
It is a good idea, maybe after the initial release we can circle up. I would definitely be interested in your thoughts on optimizations here.